File: js/src/protocol/connection.ts
/**
 * Abstract duplex byte-message channel — mirrors Python opendesk.protocol.connection.
 *
 * Concrete implementations:
 *   LoopbackConnection  — in-process pair (testing)
 *   WebSocketConnection — see transports/websocket.ts
 *   EncryptedConnection — see auth/encrypted.ts
 */

export class ConnectionClosed extends Error {
  constructor(msg = "connection closed") {
    super(msg);
    this.name = "ConnectionClosed";
  }
}

export abstract class Connection {
  /** Send a single binary message. */
  abstract send(data: Buffer): Promise<void>;
  /** Receive a single binary message.  Resolves when one arrives. */
  abstract recv(): Promise<Buffer>;
  /** Close the connection.  Idempotent. */
  abstract close(): Promise<void>;
}

// ---------------------------------------------------------------------------
// LoopbackConnection — two in-process halves
// ---------------------------------------------------------------------------

interface Waiter {
  resolve: (data: Buffer) => void;
  reject: (err: Error) => void;
}

class LoopbackHalf extends Connection {
  private queue: Buffer[] = [];
  private waiters: Waiter[] = [];
  private closed = false;
  private other!: LoopbackHalf;

  _link(other: LoopbackHalf) {
    this.other = other;
  }

  async send(data: Buffer): Promise<void> {
    if (this.closed) throw new ConnectionClosed();
    this.other._deliver(data);
  }

  _deliver(data: Buffer) {
    if (this.waiters.length > 0) {
      this.waiters.shift()!.resolve(data);
    } else {
      this.queue.push(data);
    }
  }

  async recv(): Promise<Buffer> {
    if (this.closed && this.queue.length === 0) throw new ConnectionClosed();
    if (this.queue.length > 0) return this.queue.shift()!;
    return new Promise((resolve, reject) => {
      this.waiters.push({ resolve, reject });
    });
  }

  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    const err = new ConnectionClosed("peer closed connection");
    for (const w of this.waiters) w.reject(err);
    this.waiters = [];
  }
}

export function loopbackPair(): [Connection, Connection] {
  const a = new LoopbackHalf();
  const b = new LoopbackHalf();
  a._link(b);
  b._link(a);
  return [a, b];
}

File: python/opendesk/protocol/connection.py
"""The transport-shaped interface :class:`~opendesk.protocol.peer.Peer` runs on.

A :class:`Connection` is a duplex byte-message channel.  The Peer pushes
encoded frames into ``send`` and pulls them from ``recv``.  Concrete
implementations live elsewhere:

* :class:`LoopbackConnection` — two halves connected by in-process queues,
  for hermetic testing.
* ``WebSocketConnection`` (forthcoming) — a single WebSocket binary channel.
* ``QuicConnection`` (later) — one QUIC stream per direction.

The Peer never assumes ordering across different ``send`` calls beyond
in-order delivery on a single Connection.
"""

from __future__ import annotations

import asyncio
from abc import ABC, abstractmethod
from typing import Optional


class ConnectionClosed(ConnectionError):
    """Raised when ``recv`` or ``send`` is called on a closed Connection."""


class Connection(ABC):
    """Duplex byte-message channel.  Each ``send``/``recv`` carries one frame."""

    @abstractmethod
    async def send(self, data: bytes) -> None:
        """Send a single binary message.  Raises :class:`ConnectionClosed`."""

    @abstractmethod
    async def recv(self) -> bytes:
        """Receive a single binary message.

        Blocks until a message arrives.  Raises :class:`ConnectionClosed` when
        the peer half-closes the channel.
        """

    @abstractmethod
    async def aclose(self) -> None:
        """Close the connection.  Idempotent."""

    async def __aenter__(self) -> "Connection":
        return self

    async def __aexit__(self, exc_type, exc, tb) -> None:
        await self.aclose()


class LoopbackConnection(Connection):
    """An in-process Connection — useful for testing :class:`Peer` without sockets.

    Use :meth:`pair` to get the two halves of a single duplex link::

        a, b = LoopbackConnection.pair()
        # Anything ``a.send`` sends, ``b.recv`` will receive, and vice versa.
    """

    def __init__(self, outbox: asyncio.Queue, inbox: asyncio.Queue) -> None:
        self._outbox = outbox
        self._inbox = inbox
        self._closed = False

    @classmethod
    def pair(cls) -> tuple["LoopbackConnection", "LoopbackConnection"]:
        a_to_b: asyncio.Queue = asyncio.Queue()
        b_to_a: asyncio.Queue = asyncio.Queue()
        return cls(a_to_b, b_to_a), cls(b_to_a, a_to_b)

    async def send(self, data: bytes) -> None:
        if self._closed:
            raise ConnectionClosed("send on closed connection")
        await self._outbox.put(data)

    async def recv(self) -> bytes:
        if self._closed:
            raise ConnectionClosed("recv on closed connection")
        data: Optional[bytes] = await self._inbox.get()
        if data is None:
            self._closed = True
            raise ConnectionClosed("peer closed connection")
        return data

    async def aclose(self) -> None:
        if self._closed:
            return
        self._closed = True
        try:
            self._outbox.put_nowait(None)
        except asyncio.QueueFull:  # pragma: no cover
            pass
